### PR TITLE
Add neo-beta-1 support to v14.x.x

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
(to release v14.1.x)

This adds back elementary support for neo to v14.x.x, then it rebases neo branch and finally does all
necessary changes to compile neo with swarm v5.x.x.

### CODE REVIEW

**Comment by nemanja-boric-sociomantic** [comment]
Blocked on https://github.com/sociomantic-tsunami/swarm/pull/323

Just last three commits are new, other is just rebased from `neo` branch + revert of neo removal.

File: src/dlsproto/node/neo/request/GetRange.d [hunk 1]
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
...
         Params:
-            connection = connection to client
-            msg_payload = initial message read from client to begin the request
-                (the request code and version are assumed to be extracted)
+            init_payload = initial message payload read from the connection
 
     ***************************************************************************/
 
-    final public void handle ( RequestOnConn connection, Const!(void)[] msg_payload )
+    public void preSupportedCodeSent ( Const!(void)[] msg_payload )
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: It's recommended to use the new new request handler framework, which doesn't require these separate pre and post methods. See https://github.com/sociomantic-tsunami/swarm/blob/v5.x.x/src/swarm/neo/node/ConnectionHandler.d#L161.

**Review by gavin-norman-sociomantic — CHANGES_REQUESTED** [review]
I think you also need to `verify` the rebased neo code.

**Comment by nemanja-boric-sociomantic** [comment]
Ah, `verify` is a good point, thanks!

File: src/dlsproto/node/neo/request/GetRange.d (hunk 1 above)
Comment (nemanja-boric-sociomantic) [reply src/dlsproto/node/neo/request/GetRange.d]: Err, but this hasn't been released yet - I've just tried to use it from swarm v5.0.x. Is it time for v5.1.0?

Comment (nemanja-boric-sociomantic) [reply src/dlsproto/node/neo/request/GetRange.d]: Even with v5.1.0, I can't slip the swarm dependency upgrade in v14.x.x of dlsproto :-(.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the updated swarm (so this is green now), `assert` to `verify` conversion and added release notes and a client readme.

**Comment by nemanja-boric-sociomantic** [comment]
And also a slight refactoring of the request handlers framework update commit (extracted `this.connection.event_dispatcher` to a `ed` field of a request handler).

File: src/dlsproto/node/neo/request/GetRange.d (hunk 1 above)
Comment (gavin-norman-sociomantic) [reply src/dlsproto/node/neo/request/GetRange.d]: Ahh sorry, I forgot it wasn't in v5.0.x.

File: src/dlsproto/client/README.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
+DLS Client Overview
+===================
+
+This package contains the client to connect to a Distributed Log Storage
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/README.rst]: Officially it's Store, not Storage.

File: src/dlsproto/client/README.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+The request API modules provide a single, centralised point of documentation and
+definitions pertaining to each request.
+
+Available Requests
+..................
+
+The DLS supports the following requests (links to the API modules):
+
+* `Put <request/Put.d>`_:
+  pushes a record to a channel.
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/README.rst]: Maybe "adds a record to a channel at the specified timestamp."?

File: src/dlsproto/client/README.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+* `GetRange <request/GetRange.d>`_:
+  receives a batch of records in the given range, optionally filtered by a
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/README.rst]: Something like "in the given timestamp range"?

File: src/dlsproto/node/neo/request/GetRange.d [hunk 5]
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
...
         }
         catch (Exception e)
         {
             log.error("{}", getMsg(e));
-            sendFinishedCode();
         }
-        finally
+    }
+
+    public void postSupportedCodeSent ()
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: Doc missing.

File: src/dlsproto/node/neo/request/GetRange.d [hunk 6]
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
...
+    {
+        this.value_buffer = this.resources.getVoidBuffer();
+        this.batch_buffer = this.resources.getVoidBuffer();
+        this.compressed_batch = this.resources.getVoidBuffer();
+        this.saved_exception = this.resources.getException();
+        this.lzo = this.resources.getLzo();
+
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: Double newline.

File: src/dlsproto/node/neo/request/GetRange.d [hunk 7]
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
...
         {
             log.error("{}", getMsg(e));
-            sendFinishedCode();
         }
-        finally
+    }
+
+    public void postSupportedCodeSent ()
+    {
+        this.value_buffer = this.resources.getVoidBuffer();
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/GetRange.d]: Small optimization: you could only acquire these resources if the request was initialised ok.

File: src/dlsproto/node/neo/request/Put.d [hunk 8]
```diff
@@ -12,77 +12,70 @@
...
-                ( ed.Payload payload )
-                {
-                    payload.addConstant(RequestStatusCode.Error);
-                }
-            );
+            this.response = RequestStatusCode.Error;
         }
     }
 
+    public void postSupportedCodeSent ()
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/Put.d]: Missing doc.

File: src/dlsproto/node/neo/request/Put.d [hunk 9]
```diff
@@ -12,77 +12,70 @@
...
-        this.connection = connection;
-        auto ed = this.connection.event_dispatcher;
-        auto parser = ed.message_parser;
-
         // Extract the channel, record's timestamp and value from the
         // message payload
-        cstring channel = parser.getArray!(char)(msg_payload);
-        time_t timestamp = *parser.getValue!(time_t)(msg_payload);
-        auto value = parser.getArray!(char)(msg_payload);
+        cstring channel = this.ed.message_parser.getArray!(char)(msg_payload);
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/node/neo/request/Put.d]: Btw David recommended using `parseBody`. See https://github.com/sociomantic-tsunami/dhtproto/pull/110#discussion_r187033167.

Comment (nemanja-boric-sociomantic) [reply src/dlsproto/node/neo/request/Put.d]: Nice tip, I forgot that this was there.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with all comments addressed.

**Comment by nemanja-boric-sociomantic** [comment]
Red because of sourceforge issues.

**Review by gavin-norman-sociomantic — APPROVED** [review]
LGTM. Feel free to auto-merge when CI is back up.

**Pull request merged**